### ISSUE TITLE
Extend title_format to support i3: title bars

### ISSUE THREAD

**Issue by memeplex** [i3/i3#3343]
## I'm submitting a…

<pre>
[ ] Bug
[x] Feature Request
[ ] Documentation Request
[ ] Other (Please describe in detail)
</pre>

## Current Behavior

I'm using title_format to add 1 space indentation to titles. But I see no way to match `i3: X[....]` titles since all matching criteria seems to target a proper client and title_format cannot be set without outside any criteria.

This makes vertically stacked window titles ill-aligned.

## Expected Behavior

title_format = "..." is accepted and applies to any titlebar (or only to i3: ones).

## Environment

Output of `i3 --moreversion 2>&-`:

<pre>
Binary i3 version:  4.15 (2018-03-10) © 2009 Michael Stapelberg and contributors
Running i3 version: 4.15 (2018-03-10) (pid 11418) abort…)
Loaded i3 config: /home/carlos/.config/i3/config (Last modified: Tue 24 Jul 2018 03:58:40 PM -03, 459 seconds ago)

The i3 binary you just called: /usr/bin/i3
The i3 binary you are running: i3
</pre>

Configuration file:

<pre>
# General settings
set $mod Mod4
floating_modifier $mod
font pango:DejaVu Sans Bold 10
default_border normal 1
hide_edge_borders both
workspace_auto_back_and_forth yes
set $nexec exec --no-startup-id

# Theme
set $bg #13191c
set $border #101619
set $fg #83898c
set $fg_high #cfd7d3
set $fg_urg #cc6666
#                       Border  Background Text     Indicator Child border
client.focused          $border $bg        $fg_high $fg       $border
client.focused_inactive $border $bg        $fg      $fg       $border
client.unfocused        $border $bg        $fg      $fg       $border
client.urgent           $border $bg        $fg_urg  $fg       $border
client.placeholder      $border $bg        $fg      $fg       $border
client.background       $border

# App shortcuts
bindsym $mod+l $nexec emacsclient -c
bindsym $mod+Control+l exec xfce4-terminal
bindsym $mod+Shift+l exec firefox
bindsym $mod+Mod1+l exec spotify
bindsym $mod+Tab $nexec rofi -no-fixed-num-lines -location 7 -show window -yoffset -25
bindsym $mod+Return $nexec "rofi -location 7 -combi-modi drun,run -show combi -yoffset -25"
bindsym $mod+Control+Return $nexec music toggle
bindsym $mod+Shift+Return $nexec music mute
bindsym $mod+Mod1+Return $nexec dual
bindsym $mod+Shift+Control+Return $nexec passmenu
bindsym $mod+Left $nexec music prev
bindsym $mod+Right $nexec music next
bindsym $mod+Down $nexec music down
bindsym $mod+Up $nexec music up
bindsym $mod+Shift+Down $nexec xbacklight -dec 2
bindsym $mod+Shift+Up $nexec xbacklight -inc 2
bindsym $mod+s $nexec search $(xclip -o)
bindsym $mod+Control+s $nexec search -r
bindsym $mod+o $nexec shot %c
bindsym $mod+Control+o $nexec shot %d
bindsym $mod+Shift+o $nexec shot %c 5
bindsym $mod+Mod1+o $nexec shot %d 5

# Change focus
bindsym $mod+b focus left
bindsym $mod+n focus down
bindsym $mod+Shift+Control+n focus child
bindsym $mod+p focus up
bindsym $mod+Shift+Control+p focus parent
bindsym $mod+f focus right
bindsym $mod+grave focus right

# Change layout
bindsym $mod+w layout splith
bindsym $mod+Control+w layout splitv
bindsym $mod+a layout tabbed
bindsym $mod+Control+a layout stacking
bindsym $mod+d split h
bindsym $mod+Control+d split v
bindsym $mod+m fullscreen toggle
bindsym $mod+Control+m border toggle
bindsym $mod+space focus mode_toggle
bindsym $mod+Control+space floating toggle

# Move window
bindsym $mod+Control+b move left
bindsym $mod+Control+n move down
bindsym $mod+Control+p move up
bindsym $mod+Control+f move right
bindsym $mod+Control+1 move container to workspace 1
bindsym $mod+Control+2 move container to workspace 2
bindsym $mod+Control+3 move container to workspace 3
bindsym $mod+Control+4 move container to workspace 4
bindsym $mod+Control+5 move container to workspace 5
bindsym $mod+Control+6 move container to workspace 6
bindsym $mod+Control+7 move container to workspace 7
bindsym $mod+Control+8 move container to workspace 8
bindsym $mod+Control+9 move container to workspace 9
bindsym $mod+Control+0 move container to workspace 10

# Resize window
bindsym $mod+Control+Left resize shrink width 10 px or 5 ppt
bindsym $mod+Control+Up resize grow height 10 px or 5 ppt
bindsym $mod+Control+Down resize shrink height 10 px or 5 ppt
bindsym $mod+Control+Right resize grow width 10 px or 5 ppt

# Switch to workspace
bindsym $mod+1 workspace 1
bindsym $mod+2 workspace 2
bindsym $mod+3 workspace 3
bindsym $mod+4 workspace 4
bindsym $mod+5 workspace 5
bindsym $mod+6 workspace 6
bindsym $mod+7 workspace 7
bindsym $mod+8 workspace 8
bindsym $mod+9 workspace 9
bindsym $mod+0 workspace 10

# Move workspace
bindsym $mod+Shift+1 rename workspace to 1
bindsym $mod+Shift+2 rename workspace to 2
bindsym $mod+Shift+3 rename workspace to 3
bindsym $mod+Shift+4 rename workspace to 4
bindsym $mod+Shift+5 rename workspace to 5
bindsym $mod+Shift+6 rename workspace to 6
bindsym $mod+Shift+7 rename workspace to 7
bindsym $mod+Shift+8 rename workspace to 8
bindsym $mod+Shift+9 rename workspace to 9
bindsym $mod+Shift+0 rename workspace to 10
bindsym $mod+Mod1+p move workspace to output up
bindsym $mod+Mod1+n move workspace to output down
bindsym $mod+Mod1+f move workspace to output right
bindsym $mod+Mod1+b move workspace to output left

# Marks
bindsym $mod+Control+F1 mark --toggle 1
bindsym $mod+F1 [con_mark=1] focus
bindsym $mod+Control+F2 mark --toggle 2
bindsym $mod+F2 [con_mark=2] focus
bindsym $mod+Control+x mark --toggle x
bindsym $mod+Shift+x move window to mark x, unmark x
bindsym $mod+x swap container with mark x, unmark x

# Scratchpad
bindsym $mod+z scratchpad show
bindsym $mod+Control+z floating enable, move scratchpad
bindsym $mod+Shift+z scratchpad show, scratchpad show
bindsym $mod+e [title="^Scratchpad - Emacs"] scratchpad show
bindsym $mod+Control+e [title="^Scratchpad - Terminal$"] scratchpad show
bindsym $mod+Shift+e [title="(Gmail|Slack|WhatsApp)"] focus
bindsym $mod+Mod1+e [class="^Spotify$"] scratchpad show
bindsym $mod+Shift+Control+e $nexec emacsclient -F '((title . "Scratchpad - Emacs"))' -a '' -c ~/notes/inbox.org
bindsym $mod+Shift+Control+Mod1+e exec xfce4-terminal --title 'Scratchpad - Terminal'

# Control
bindsym $mod+r exec i3-input
bindsym $mod+Control+r reload
bindsym $mod+Shift+r restart
bindsym $mod+Shift+Control+r exit
bindsym $mod+c exec i3lock -c 1d1f21
bindsym $mod+Control+c exec systemctl sleep
bindsym $mod+Shift+c exec systemctl hibernate
bindsym $mod+Shift+Control+c exec systemctl poweroff
bindsym $mod+Control+q kill

# Rules
for_window [class="^Pavucontrol$"] floating enable
for_window [class="^qpaeq$"] floating enable
for_window [class="^Gpick$"] floating enable
for_window [class="^Connman-gtk$"] floating enable
for_window [class="^Xmag$"] floating enable
for_window [class="^Spotify$"] floating enable, resize set 1000 700, move position center
for_window [title="^Htop$"] floating enable
for_window [title="^Scratchpad - Terminal$"] floating enable, move scratchpad
for_window [title="^Scratchpad - Emacs$"] floating enable, move scratchpad

# Startup
$nexec emacsclient -F '((title . "Scratchpad - Emacs"))' -a '' -c ~/notes/inbox.org
$nexec xfce4-terminal --title 'Scratchpad - Terminal'
</pre>

<pre>
- Linux Distribution & Version: Linux archlinux 4.17.8-1-ARCH #1 SMP PREEMPT Wed Jul 18 09:56:24 UTC 2018 x86_64 GNU/Linux
- Are you using a compositor: compton
</pre>

**Comment by orestisf1993** [comment]
I think the problem here is that `for_window` will only match leaf containers (windows), not i3 frames.

I understand the use case though.

**Comment by memeplex** [comment]
Since there is not much else to set for a container than its title format maybe allowing standalone `title_format` with the meaning "applies to everything" will suffice. Then the user can override it using rules.

**Comment by orestisf1993** [comment]
This would not be backwards compatible with the current `title_format`. It could work if it was a config directive instead of a command but it would mean having the same name for a config directive and a command.

**Comment by memeplex** [comment]
What about default_title_format as in default_border?

**Comment by orestisf1993** [comment]
@Airblader

**Comment by Airblader** [comment]
I don't think `default_title_format` solves the underlying problem here. It just means that eventually someone will want the title_format of those containers to be different on one of the workspaces.

On a side note, using `[con_id=…] title_format` works using the con_id of the container window.

**Comment by memeplex** [comment]
You're right. Besides, I now realize that parent containers can legitimately be made floatings, moved to the scratchpad, etc. So extending rules to match them seems to be the right approach. Moreover, if any operation is not supported by an inner container it can nevertheless be applied now by just selecting the container by hand and running the problematic command, so these cases should already have been addressed at this level (by signalling an error or whatever) and it wouldn't be necessary a case by case decision while extending rules to i3 containers. Am I wrong?

**Comment by Airblader** [comment]
We cannot just extend for_window to match these containers because it would be a breaking change. And lots of people would probably want to do the opposite and restrict to leaf containers. 

So it'd have to be an additional criterion similar to floating / tiling, and its absence matches only leaf containers. 

However, the question is how useful this is. There's really never anything specific about these containers that is worth matching on (other than matching all of these, specifically). So that's why I'm not an immediate fan of adding a criterion for it. 

Perhaps a separate directive which doesn't accept match criteria would be better?

**Comment by memeplex** [comment]
> Perhaps a separate directive which doesn't accept match criteria would be better?

I'm not quite following you here. My suggestion about `default_title_format` was pointing in that direction, but you persuasively argued against it. So what kind of separate directive that doesn't accept matching criteria would that be?

**Comment by Airblader** [comment]
The problem with default_title_format is that it already defines the property to set. We don't want to prevent flexibility as to what to execute, just remove useless flexibility as to what to match.

So what I'm referencing would be something like `for_container`, just without criteria:

    for_container title_format ... 
    for_container border pixel 5

(Pleaee don't take this as a decision, so far that's just my thoughts and a proposal)

**Comment by memeplex** [comment]
I understand but there are two different axes for extensibility here:

1. The one you pointed out: the user could want to apply different formats to, say, different workspaces. This is the criteria axis.

2. The one I pointed out once I acknowledged the shortcomings of my initial proposal: the user might want to do many other things with inner containers, make them float, scratch them, etc. This is the action axis.

Now, default_title_format allows extensibility in neither axis, but for_container seems to be precluding extensibility along the criteria axis. So what if the user wants to match a particular workspace?

**Comment by Airblader** [comment]
We'd have to consider whether we'd even want to allow this, but if we did, the syntax could be extended to optionally allow criteria which is a non-breaking change.

On the other hand, default_title_format couldn't be extended in any sensible way to allow different commands. 

On a side note, I'm not happy with the name for_container. It's just a placeholder.

**Comment by memeplex** [comment]
But you do prefer `for_container [...]` over `for_window [..., containers=true]`, don't you? Why?

**Comment by Airblader** [comment]
Because the containers we're talking about aren't windows, so for_window seems like a misnomer for this. What we lose is the ability to have a single line to affect both. However, I think that's an acceptable limitation, especially assuming very few users would want to do this in the first place.

**Comment by memeplex** [comment]
Ok, it also allows for a more gradual implementation if you don't want to support full blown criteria at first.

**Comment by memeplex** [comment]
Btw, this seems to work quite well as a workaround using the ipc interface from python:

```python
import i3ipc

i3 = i3ipc.Connection()

def on_window_new(i3, ev):
    for con in i3.get_tree():
        if con.type == 'con' and con.nodes:
            con.command('title_format " %title"')

i3.on('window::new', on_window_new)

i3.main()
```

Not that I like it too much...

**Comment by memeplex** [comment]
Now that I have some of the most i3 savvy people here, is there any better way to match container creation than that above? There I'm catching window creation events and then visiting the entire tree to update containers that were possibly created for the sake of the new window.

**Comment by Airblader** [comment]
I don't think we have events for that, currently, but I'd be open to adding something like `container::open` (and possibly move + close, if necessary). I'd also prefer that over adding the config directive (for now) and think the work going into it would be similarly complex anyway.

**Comment by memeplex** [comment]
@Airblader today I ran into a new use case for which having the container::open/new/whatever API would have made a big difference: I wanted to dynamically set the layout of a new container to tabbed in a small screen and to splith in a big screen. Again, I ended up trapping new window events and checking whether the window in questionwas the first and only child of its container or not, etc. This is not robust, nor simple, nor efficient. I came to believe that the API is really lacking in this regard.

**Comment by ismig** [comment]
Any chance a `for_container` will be added?
I'd be happy with a flag like `shorten_container_titles yes` to make titles less noisy. E.g., `i3: T[google-chrome google-chrome]` could be shortened `tabbed (2)`. The sub-container count would be more useful to me than the current sub-container titles.
